src/thread.rs

```diff
@@ -1,14 +1,82 @@
 ///! TODO
-use std::{ffi::CStr, ptr::NonNull};
-
-use crate::libspdk::{spdk_get_thread, spdk_thread, spdk_thread_get_name};
+use crate::cpu_cores::Cores;
+use crate::{
+    cpu_cores::CpuMask,
+    libspdk::{
+        spdk_get_thread,
+        spdk_set_thread,
+        spdk_thread,
+        spdk_thread_create,
+        spdk_thread_destroy,
+        spdk_thread_exit,
+        spdk_thread_get_by_id,
+        spdk_thread_get_id,
+        spdk_thread_get_name,
+        spdk_thread_is_exited,
+        spdk_thread_poll,
+        spdk_thread_send_msg,
+    },
+};
+use futures::channel::oneshot::{channel, Receiver, Sender};
+use nix::{errno::Errno, libc};
+use std::{
+    ffi::{c_void, CStr, CString},
+    fmt::Debug,
+    future::Future,
+    ptr::NonNull,
+};
 
 /// Wrapper for `spdk_thread`.
+#[derive(Debug, PartialEq, Clone, Copy)]
 pub struct Thread {
     inner: NonNull<spdk_thread>,
 }
 
+unsafe impl Send for Thread {}
+
 impl Thread {
+    /// With the given thread as context, execute the closure on that thread.
+    ///
+    /// Any function can be executed here however, this should typically be used
+    /// to execute functions that reference any FFI to SPDK.
+    pub fn new(name: String, core: u32) -> Option<Self> {
+        let name = CString::new(name).unwrap();
+
+        NonNull::new(unsafe {
+            let mut mask = CpuMask::new();
+            mask.set_cpu(core, true);
+            spdk_thread_create(name.as_ptr(), mask.as_ptr())
+        })
+        .map(|inner| Self {
+            inner,
+        })
+    }
+
+    /// Destroys the given thread waiting for it to become ready to destroy.
+    pub fn destroy(self) {
+        debug!(
+            "Destroying SPDK thread {}... {:p}",
+            self.name(),
+            self.as_ptr()
+        );
+
+        unsafe {
+            spdk_set_thread(self.as_ptr());
+
+            // set that we *want* to exit, but we have not exited yet
+            spdk_thread_exit(self.as_ptr());
+
+            // now wait until the thread is actually exited the internal
+            // state is updated by spdk_thread_poll()
+            while !spdk_thread_is_exited(self.as_ptr()) {
+                spdk_thread_poll(self.as_ptr(), 0, 0);
```

Comment:
I probably don't have the right context here, but could this be starving our "async" reactors?

Reply:
This is our code from mayastor, I just moved it here as is to clean up code structure. I cannot tell it is really causes any problems to async reactors or not.